### ISSUE TITLE
[MINVOKER-209]  ITs fail (on Windows 10) when working directory contains space

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-invoker-plugin#612]
**[Alix Lourme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=axel3rd)** opened **[MINVOKER-209](https://issues.apache.org/jira/browse/MINVOKER-209?redirect=false)** and commented

When working copy contains some space (common situation on W10 with users directories Firstname/Lastname), most of ITs fail.

### Scenario

Command:

```
mvn invoker:run -Dinvoker.test=special-characters-should-work -Dfile.encoding=UTF-8 -Prun-its,dev
```

#### working directory with spaces

_build.log_ result :

```
Running pre-build script: C:\Users\Firstname Lastname\git\maven-plugins\maven-invoker-plugin\target\it\special-characters-should-work\setup.groovy
Finished pre-build script: C:\Users\Firstname Lastname\git\maven-plugins\maven-invoker-plugin\target\it\special-characters-should-work\setup.groovy
Erreur : impossible de trouver ou charger la classe principale Lastname\git\maven-plugins\maven-invoker-plugin\target
```

Content of : _DefaultInvoker.executeCommandLine(Commandline, InvocationRequest)_ (line: ~126)

cmd:

```
cmd.exe /X /C ""C:\Program Files\[MavenDirectory]\bin\mvn.cmd" -B -X -D "maven.repo.local=C:\Users\Firstname Lastname\git\maven-plugins\maven-invoker-plugin\target\local-repo" -s "C:\Users\Firstname Lastname\git\maven-plugins\maven-invoker-plugin\target\it\interpolated-settings.xml" -D maven.compiler.source=1.6 -D maven.compiler.target=1.6 clean verify"
```

workDir:

```
C:\Users\Firstname Lastname\git\maven-plugins\maven-invoker-plugin\target\it\special-characters-should-work
```

### working directory without space

_build.log_ result:

```
Running pre-build script: c:\Windows\Temp\maven-invoker-plugin\target\it\special-characters-should-work\setup.groovy
Finished pre-build script: c:\Windows\Temp\maven-invoker-plugin\target\it\special-characters-should-work\setup.groovy
Apache Maven 3.4.0-SNAPSHOT (54827e6b0fc02425f701cfcccde46f81edcb7058; 2016-10-01T18:01:52+02:00)
... next lines are ok : BUILD SUCCESS
```

Content of :  _DefaultInvoker.executeCommandLine(Commandline, InvocationRequest)_ (line: ~126)

cmd:

```
cmd.exe /X /C ""C:\Program Files\[MavenDirectory]\bin\mvn.cmd" -B -X -D maven.repo.local=C:\Windows\Temp\maven-invoker-plugin\target\local-repo -s C:\Windows\Temp\maven-invoker-plugin\target\it\interpolated-settings.xml -D maven.compiler.source=1.6 -D maven.compiler.target=1.6 clean verify"
```

workDir:

```
C:\Windows\Temp\maven-invoker-plugin\target\it\special-characters-should-work
```

I'm not sure of the root cause (invoker initialization or _plexus-utils_ problem :()

---

**Affects:** 3.0.0

**Attachments:**
- [MINVOKER-209-maven-invoker-plugin-IT-double-exec-special-characters-should-work.patch](https://issues.apache.org/jira/secure/attachment/12836251/MINVOKER-209-maven-invoker-plugin-IT-double-exec-special-characters-should-work.patch) (_861 bytes_)
- [MINVOKER-209-workaround-maven-invoker-plugin.patch](https://issues.apache.org/jira/secure/attachment/12836993/MINVOKER-209-workaround-maven-invoker-plugin.patch) (_5.21 kB_)

**Issue Links:**
- [MNG-4559](https://issues.apache.org/jira/browse/MNG-4559) MAVEN_OPTS are incorrectly resolved in Unix
 (_**"is blocked by"**_)
- [MINVOKER-211](https://issues.apache.org/jira/browse/MINVOKER-211) Some ITs can't be re-executed without clean

- [MNG-5962](https://issues.apache.org/jira/browse/MNG-5962) mvn.cmd fails when the current directory has spaces in between
 (_**"depends upon"**_)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

This is the same workaround I had in my mind, though I would really like to see a real fix for that. Please create a patch, I will validate and apply it.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Thanks, I will have a look at the patch later that day.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [r1768121](http://svn.apache.org/r1768121). Passes on Windows 10 and FreeBSD 10.3-RELEASE. I have applied the patch to `pom.xml` only. Please create a separate issue for the remaining improvements.

**Comment by jira-importer** [comment]
**[Alix Lourme](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=axel3rd)** commented

> Please create a separate issue for the remaining improvements.
> Please consider MINVOKER-211

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Reopening, IT 'special-characters-should-work' is still failing on my system. Traced it back to r1763932 where the test was adjusted with extra spaces.

```
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T17:41:47+01:00)
Maven home: d:\apache-maven-3.3.9\bin\..
Java version: 1.7.0_55, vendor: Oracle Corporation
Java home: C:\Program Files\Java\jdk1.7.0_55\jre
Default locale: nl_NL, platform encoding: Cp1252
OS name: "windows 8.1", version: "6.3", arch: "amd64", family: "windows"
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Did you retest with Maven's pre-reset master branch?

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

No, but if this means that it doesn't work for Windows + M3.3.9 or less, that should be added to IT as criteria.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Confirmed on:

```
D:\Entwicklung\Projekte\maven invoker plugin>mvn -v
Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T17:41:47+01:00)
Maven home: D:\Entwicklung\Programme\apache-maven-3.3.9\bin\..
Java version: 1.8.0_112, vendor: Oracle Corporation
Java home: C:\Program Files\Java\jdk1.8.0_112\jre
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "dos"
```

with

```
[INFO] Building: special-characters-should-work\pom.xml
[INFO] run script setup.groovy
[INFO] ..FAILED (3.0 s)
[INFO]   The build exited with code 1. See D:\Entwicklung\Projekte\maven invoker plugin\target\it\special-characters-should-work\build.log for details.
[INFO] Building: staging-dependencies\pom.xml
[INFO] run script verify.bsh
[INFO] ..SUCCESS (5.2 s)
[INFO] Building: staging-pom\pom.xml
[INFO] run script verify.bsh
[INFO] ..SUCCESS (2.0 s)
[INFO] Building: staging-reactor\pom.xml
[INFO] run script verify.bsh
[INFO] ..SUCCESS (3.4 s)
[INFO] Building: wrongPropertyInterpolation\pom.xml
[INFO] ..SUCCESS (12.1 s)
[INFO]
[INFO] --- maven-checkstyle-plugin:2.15:check (checkstyle-check) @ maven-invoker-plugin ---
[INFO] There is 1 error reported by Checkstyle 6.1.1 with config/maven_checks.xml ruleset.
[INFO] Ignored 1 error, 0 violation remaining.
[INFO]
[INFO] --- maven-invoker-plugin:2.0.0:verify (integration-test) @ maven-invoker-plugin ---
[INFO] -------------------------------------------------
[INFO] Build Summary:
[INFO]   Passed: 56, Failed: 1, Errors: 0, Skipped: 0
[INFO] -------------------------------------------------
[ERROR] The following builds failed:
[ERROR] *  special-characters-should-work\pom.xml
[INFO] -------------------------------------------------
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 04:33 min
[INFO] Finished at: 2017-01-16T20:21:19+01:00
[INFO] Final Memory: 61M/1339M
[INFO] ------------------------------------------------------------------------
```

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

For now I'd suggest to revert the changes r1763932 which added spaces to the IT (the special characters is already a valuable test). Next create 2 ITs for the spaces: one for non-windows, one for windows + M3.5.0+ and a big fat note saying these can be combined once MINVOKER-214 is available.
[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) I can pick this up, just let me know.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Yes, please go ahead and do so. Thank you!

**Comment by jira-importer** [comment]
**[Robert Scholte](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rfscholte)** commented

Fixed in [r1779169](http://svn.apache.org/viewvc?rev=1779169&view=rev) (Hudson fails due to m-javadoc-p)
